bundles/org.openhab.binding.modbus.foxinverter/src/main/java/org/openhab/binding/modbus/foxinverter/internal/MQ2200InverterHandler.java

```diff
@@ -0,0 +1,434 @@
+                                });
+                    }, (AsyncModbusFailure<ModbusReadRequestBlueprint> error) -> {
+                        // reading properties is not critical, just log at debug level
+                        logger.debug("Reading rated/max active power failed", error.getCause());
+                    });
+            // wait for async polls to complete; waiting is not ideal but not worth implementing a complex logic here
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            logger.debug("Interrupted while reading device properties");
+            Thread.currentThread().interrupt();
+        }
+        return properties;
+    }
+
+    private void processAlarmState() {
+        boolean currentAlarmState = (alarm[0] != 0) || (alarm[1] != 0) || (alarm[2] != 0) || statusFault;
+        if (currentAlarmState != alarmState) {
+            alarmState = currentAlarmState;
+            if (alarmState) {
+                logger.warn("Inverter {} is in Alarm State: Alarm1=0x{}, Alarm2=0x{}, Alarm3=0x{}, StatusFault={}",
+                        getThing().getUID(), Integer.toHexString(alarm[0]), Integer.toHexString(alarm[1]),
+                        Integer.toHexString(alarm[2]), statusFault);
+                // indicate severe error also by setting the ThingStatus
+                // (UNKNOWN seems more appropriate than OFFLINE, as the device still gets commands)
+                updateStatus(ThingStatus.UNKNOWN);
+            } else {
+                logger.info("Inverter {}, alarm resolved", getThing().getUID());
+                updateStatus(ThingStatus.ONLINE);
+            }
+        }
+        OpenClosedType state = alarmState ? OpenClosedType.CLOSED : OpenClosedType.OPEN;
+        logger.debug("STATUS_ALARM {} -> {}", alarmState, state);
+        updateState(new ChannelUID(thing.getUID(), "fi-overview", "fi-status-alarm"), state);
+    }
+
+    private void processHiddenChannel(MQ2200InverterRegisters channel, org.openhab.core.types.State v) {
+        // this block deals with channels which are not directly exposed, but
+        // used to update other channels
+        logger.trace("Update on internal channel {} to {}", channel.getChannelName(), v);
+        DecimalType d = v.as(DecimalType.class);
+        if (d == null) {
+            logger.warn("Internal channel {} is not DecimalType, cannot process", channel.getChannelName());
+        } else {
+            int i = d.intValue();
+            switch (channel.getChannelName()) {
+                // active power reports -30000 W during startup, to be suppressed
+                case "hidden-active-power":
+                    if (i != -30000) {
+                        logger.debug("{} {}", "ACTIVE_POWER", v);
+                        updateState(
+                                new ChannelUID(thing.getUID(), "fi-" + channel.getChannelGroup(), "fi-active-power"),
```

Comment:
Can you use the channel id's as constant instead of magic strings.
Applies to other places too.

Reply:
I am not sure how to proceed here. My channel ID is `hidden-active-power` - if I get it correctly. Should I define constants somewhere? Or do I overlook where to get it from?

Reply:
I have now changed mostly everything to use constants. I don't think that constants for all channels are needed, as there is some enum magic in InverterRegisters.java (but those enum constants/their anem function cannot be used in switch statement).